Implement new claim API and do minor updates

ed255 opened the pull request:
Resolve #296

The new Claim API works with a new Metadata struct that allows setting the
following properties of a claim in a generic way:
- Optional Recipient Identity
- Optional Version
- Optional Expiration date
- Revocation Nonce

Every Claim must contain the metadata object.  The new claims.EntrierMetadata
interface allows setting metadata properties from functions that consume claims
in a generic way.  A claim header must be defined for every claim type to
specify the flags that the claim type uses.  For now only ClaimBasic and
ClaimAuthorizeKSignBabyJub have been defined.

Other updates:
- Remove idenmanager (deprecated)
- Remove idenadminutils (deprecated)
- Remove deprecated functions from `core/proof`
- Add helper script to generate godoc links
- Update README with godoc links

arnaucube commented on core/claims/claim.go:
```diff
@@ -131,35 +111,187 @@ func NewClaimFromEntry(e *merkletree.Entry) (merkletree.Entrier, error) {
+		binary.BigEndian.PutUint32(index[0][ClaimTypeLen+ClaimFlagsLen:], m.Version)
+	}
+	if m.header.Expiration {
+		binary.BigEndian.PutUint64(value[0][32/8:], uint64(m.Expiration))
```
Why ClaimExpirationLen is a hardcoded value, while ClaimTypeLen,ClaimVersionLen, ClaimFlagsLen, ClaimFlagsLen, ClaimHeaderLen, ClaimVersionLen, ClaimRevNonceLen are declared as constants? In this way when we need to have a claim that needs Expiration, we have the constant of the len to do the parser.
With the ClaimExpirationLen or not, maybe here should use `value[0][ClaimRevNonceLen:]`?

ed255 replied:
I have modified the code to use the `ClaimRevNonceLen` constant

arnaucube commented on core/claims/claim.go:
```diff
@@ -131,35 +111,187 @@ func NewClaimFromEntry(e *merkletree.Entry) (merkletree.Entrier, error) {
+
+// EntrierMetadata is an intefrace that extends Entrier with a function that
+// returns the claim metadata.
+type EntrierMetadata interface {
```
Proposal: `EntrierMetadata` --> `Claimer`.
The idea is that a `Claimer` interface is what all the `Claims` are satisfying. So the interface don't get a name like `EntrierMetadata`, when is a superset of `Entrier`, and is more natural as superset as the `Claimer` contain a `Entrier`.

arnaucube commented on core/claims/claim.go:
```diff
@@ -131,35 +111,187 @@ func NewClaimFromEntry(e *merkletree.Entry) (merkletree.Entrier, error) {
+		m.Version = binary.BigEndian.Uint32(index[0][ClaimTypeLen+ClaimFlagsLen:])
+	}
+	if m.header.Expiration {
+		m.Expiration = int64(binary.BigEndian.Uint64(value[0][32/8:]))
```
Same as previous comment

ed255 replied:
I have modified the code to use the `ClaimRevNonceLen` constant

arnaucube commented on core/claims/claimBasic.go:
```diff
@@ -1,85 +1,70 @@
-	// 56+248+248+248=800 bits
-	indexSlotBits  = 800
+	// 152+248+248+248=896 bits
+	indexSlotBits  = 896
```
this shouldn't be 120+248+248+248= 864?

ed255 replied:
I used 152 because the version field is not being used in this claim.

arnaucube replied:
Right, perfect

arnaucube commented on core/claims/claimBasic.go:
```diff
@@ -1,85 +1,70 @@
 	// 216+248+248+248=960 bits
-	dataSlotBits  = 960
-	DataSlotBytes = dataSlotBits / 8
+	valueSlotBits  = 960
```
this shouldn't be 152+248+248+248=896?

Basically we moved 64 bits from Index to Value, so Index was 800 bits and Value was 960, moving those 64 bits from I to V, results with 864 bits for Index, and 896 bits for Value, witch matches with the numbers calculated in the review comments.
See the Claims specification docs for more details

ed255 replied:
I used 216 because the expiration date is not being used in this claim.

arnaucube replied:
Right, perfect

ed255 commented on core/claims/claim.go:
```diff
@@ -131,35 +111,187 @@ func NewClaimFromEntry(e *merkletree.Entry) (merkletree.Entrier, error) {
+
+// EntrierMetadata is an intefrace that extends Entrier with a function that
+// returns the claim metadata.
+type EntrierMetadata interface {
```
I like this proposal!

ed255 replied:
Done!

ed255 commented:
I have addressed the requested changes.  I've also changed the declaration of `IndexSlotBytes` and `ValueSlotLen` to be more explicit about how they are being calculated.  The values remain the same.
